compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs

```diff
@@ -52,16 +46,27 @@ impl<'tcx> FunctionCoverage<'tcx> {
         is_used: bool,
     ) -> Self {
         let num_counters = function_coverage_info.num_counters;
-        let num_expressions = function_coverage_info.num_expressions;
+        let num_expressions = function_coverage_info.expressions.len();
         debug!(
             "FunctionCoverage::create(instance={instance:?}) has \
             num_counters={num_counters}, num_expressions={num_expressions}, is_used={is_used}"
         );
+
+        // Every expression that has an associated mapping should have a
+        // corresponding statement in MIR. If we don't see that statement, we
+        // know that it was removed by MIR optimizations.
+        let mut expressions_not_seen = BitSet::new_empty(num_expressions);
+        for Mapping { term, .. } in &function_coverage_info.mappings {
+            if let &CovTerm::Expression(id) = term {
+                expressions_not_seen.insert(id);
```

Comment:
When this set eventually gets used (in `identify_zero_expressions`), we only want it to contain expression IDs that we *expected* to see in MIR (as `ExpressionUsed`), but didn't actually see.

So in this initialization step, we identify the subset of expression IDs that we expect to see (i.e. those that are directly used by one or more regions). During MIR traversal, every `ExpressionUsed` causes its ID to be removed from the set.

Expression IDs that are only used indirectly (by other expressions) don't have a corresponding `ExpressionUsed` statement. If we only tracked “seen” expressions, those indirect expressions would be treated as “unseen”, which isn't what we want, so we have to exempt them somewhere.

So by using upside-down logic for the set, the code that actually uses the set gets to be simpler.

(I'll try to find ways to make this more obvious.)

Reply:
I ended up deciding to reverse this and move to `expressions_seen`, because that reduced the number of double-negatives I have to use in explaining how any of this works.